### PR TITLE
optimize coarse layers separately

### PR DESCRIPTION
**Pull request by mikowals**
As mentioned in issue #1, I separate the gradient calculations and keep the hint loss out of slim.losses so the hint loss could be applied on the the coarse layers only. 

It trains reasonably well getting an error rate of 3.2% in 40 epochs which is below the coarse model in paper but not approaching the 1.3% of the full DCN.  I am rerunning with the learning rate dropping much faster as a learning rate of 0.0001 got the model to drop below 3.8% error.  

A few other changes:
- I also updated code for weight decay losses as slim adds them to tf.GraphKeys.REGULARIZATION.LOSSES not losses.LOSSES_COLLECTION. 
- I divide the hint loss by batch size so it is the same scale as the cross entropy loss.  The paper does not state that either is divided by batch but certainly cross entropy loss per example is standard practice to avoid fiddling with learning rate per batch size.  This still seems to give a squared difference between coarse and fine features of ~300 without trying to minimize the hint loss.  The comparable number from figure 2 in the paper is 25 dropping to 10 so maybe it also needs to be reduced by the number of patches.  
- I am weighting the hint loss as discussed in #1 as without a low weight the hint loss drops near 0. The paper shows a minimum of 5.0 given in figure 2 when hint loss used for training.  Even weighting by 0.0001 the squared difference across patches goes down to 2.0-3.0.  So between this and the bullet above something still seems off with the hint loss though the formula in the paper is very straightforward.  I have run out of ideas of where the difference might be coming in.

### CODE REVIEW

**Comment by beopst** [comment]
Thank you for your contribution! 

For the first bullet, slim also adds all losses defined in losses.py into slim.losses.LOSSES_COLLECTION. Therefore, we don't need to consider tf.GraphKeys.REGULARIZATION_LOSSES. If total_loss is a sum of slim losses and tf's regularization losses, regularization terms will be duplicated. 

Currently, I've been trying to reproduce DCN result without hint objective. Weird thing is that DCN without hints gives higher error rate (3.3~3.5%) than Coarse Model (applying only coarse_layers + top_layers) in my experiment. I got 2.1~2.2 % of error rate in case of Coarse Model, and 0.95~1% of error rate in case of Fine Model. Do you have any scores from Coarse and Fine Models?

If DCN without hints does not outperform Coarse Model, I think I should investigate the backpropagation pass.

**Comment by beopst** [comment]
I fixed very critical indexing mistakes in replace_features function. Now, DCN without hints outperforms Coarse Model, but its error rate is still little bit higher than Fine Model.
